### ISSUE TITLE
Use Py4j to communicate between Python and Client.scala

### ISSUE THREAD

**Issue by NielsZeilemaker** [blaze/rambling#50]
Py4j is used in PySpark to communicate between the python process and the yarn client. I think a similar setup would provide a lot of benefits for knit.

My suggestion is to remove the argument parsing, and instead start a gateway server here:
https://github.com/dask/knit/blob/master/knit_jvm/src/main/scala/io/continuum/knit/Client.scala#L35

This would be a combination of:
https://github.com/apache/spark/blob/d83c2f9f0b08d6d5d369d9fae04cdb15448e7f0d/core/src/main/scala/org/apache/spark/api/python/PythonGatewayServer.scala
and
https://www.py4j.org/

Basically registering the Client object, as the entry point. And thereby, allowing python to simply call a method in Client to start the AM/set the number of cores etc.

Similar to PySpark we'll need a socket to communicate the port of the gateway server over. Which should be similar to https://github.com/apache/spark/blob/d83c2f9f0b08d6d5d369d9fae04cdb15448e7f0d/python/pyspark/java_gateway.py

After this is done, we can start thinking of implementing #40, as we'll be able to communicate to the yarn client/rm ~directly.

@quasiben what do you think?

**Comment by quasiben** [comment]
Definitely agree.  I had similar thoughts -- adding more CLI options is annoying to say the least.  I've experimented with this a bit in the past:  https://github.com/ContinuumIO/nutchpy/tree/master/nutchpy  but was a little less diligent with tests then

We'll also want higher bandwidth communication between Client and AM especially for things like dynamic container management (ongoing discussion here: https://github.com/dask/distributed/issues/128)

My initial thoughts were also Py4J but do you think it's worth considering other options ? Jepp/JPype or Thrift/Protobuf ? 

More extensive thoughts on scala/python interaction:
https://github.com/A-OK/Snakes-and-Ladders/

**Comment by NielsZeilemaker** [comment]
I was thinking to use PY4j solely because spark is also using it. I guess it should be fast enough for dynamic allocation, but sending large binary objects is probably not a good idea.

Spark has a similar comment:
https://cwiki.apache.org/confluence/display/SPARK/PySpark+Internals
"large data transfers are performed through a different mechanism."

As for the other options, I guess Thrift stands out, as that "protocol" is used to communicate between the other yarn stuffs (AM -> RM etc.). But then again, I guess simply starting another PY4j gateway in the AM, and registering its rpc port using

```
rmClient.registerApplicationMaster("", PY4J gateway port, "")
```

would allow us to connect from python directly to the AM.

We do need to bind to 0.0.0.0 on the AM, and have some way of "securing it", as described here
https://www.py4j.org/faq.html#are-there-any-security-concerns-with-py4j.

**Comment by NielsZeilemaker** [comment]
@quasiben i've implemented a poc at #51

**Comment by koverholt** [comment]
@NielsZeilemaker, this is great! Thanks for your contribution/PR here. I just wanted to let you know that @quasiben is out of the office this week, but we can review this when he gets back.

**Comment by NielsZeilemaker** [comment]
Sure no problem, I've had a lot of fun trying this stuff out. For now, I'll leave it as it is and wait for @quasiben to return.

**Comment by martindurant** [comment]
This looks fixed t me - OK to close?

**Comment by quasiben** [comment]
Oh, yes!  Thanks for cleaning up.
